### ISSUE TITLE
Deactivating account leaves the account "stuck" in rooms

### ISSUE THREAD

**Issue by brycied00d** [matrix-org/synapse#1853]
When I setup my own homeserver, I deactivated my matrix.org account. I expected that deactivating the account would be akin to deleting it, including removing the account from the various rooms it's in. Unfortunately, this is not the case, and worse still is that my prior account had power=100 in a couple of rooms so I'm unable to `/kick` it (having power=100 with my "new" account).

I believe the solution is for the `/deactivate` endpoint(s) to also "clean up," and for there to be a periodic (or at least one-time) process to clean up deactivated accounts from rooms.

What are the developers' thoughts on this deactivation process?

**Comment by brycied00d** [comment]
I should add that, had I realized that deactivating the user would leave it "stuck," I would have "cleaned up" myself before deactivating. This speaks to a client usability issue, true, however I believe still that `/deactivate` should also perform this type of cleanup.

**Comment by onePinger** [comment]
Not sure if "stuck" addresses this, but when you deactivate your account, enter your password and click on "Deactivate Account", resulting message that the session has been signed out and clicking on "OK" returns you to the deactivate page again.

**Comment by 4nd3r** [comment]
i made script which "removes" users from rooms when they are not found in ldap.
run as root. tested on debian stretch. postgres username and dbname is "synapse".
script expects that file `/root/.pgpass` exists and ldap-utils are configured (`/etc/ldap/ldap.conf`).
`curl` and `jq` are also needed.
by default script does nothing (dry run). add `--doit` to do it for real.
**no warranties, no support, use at your own risk.**

```
#!/bin/sh -e

if [ "$2" != '--doit' ]
then echo 'DRY RUN!'
fi

psql -qtA -h localhost -U synapse -c \
'select name from users where password_hash is null and appservice_id is null;' \
| sed -nr 's/@(.+):.+/\1/p' \
| while read -r uid
do
    if ldapsearch -x "uid=$uid" dn | grep -q ^dn:
    then continue
    fi

    token="$( psql -qtA -h localhost -U synapse -c \
        "select token from access_tokens where user_id like '@$uid:%' limit 1;" )"

    if [ -z "$token" ]
    then
        echo "user $uid has no token"
        continue
    fi

    rooms="$( curl -s -o /dev/null \
        "http://localhost:8008/_matrix/client/r0/joined_rooms?access_token=$token" \
        | jq -r '.joined_rooms[]' )"

    if [ -z "$rooms" ]
    then
        sleep 1
        continue
    fi

    echo "user $uid"

    echo "$rooms" | while read -r room
    do
        echo "leave $room"
        [ "$2" == '--doit' ] && \
            curl -s -o /dev/null -X POST \
            "http://localhost:8008/_matrix/client/r0/rooms/$room/leave?access_token=$token"
        sleep 1
    done
done
```

**Comment by tulir** [comment]
This seems to have been fixed in #3201
